File: src/main/java/no/nav/sosialhjelp/soknad/business/service/oppsummering/steg/inntektformue/AndreInntekter.java
package no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.inntektformue;

import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Avsnitt;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Felt;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Sporsmal;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.oppsummering.dto.Type;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static java.util.Collections.singletonList;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.BEKREFTELSE_UTBETALING;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_ANNET;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_FORSIKRING;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_SALG;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_UTBYTTE;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.StegUtils.booleanVerdiFelt;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.inntektformue.InntektFormueUtils.harBekreftelse;
import static no.nav.sosialhjelp.soknad.business.service.oppsummering.steg.inntektformue.InntektFormueUtils.harBekreftelseTrue;

public class AndreInntekter {

    public Avsnitt getAvsnitt(JsonOkonomiopplysninger opplysninger) {
        return new Avsnitt.Builder()
                .withTittel("inntekt.inntekter.titel")
                .withSporsmal(andreInntekterSporsmal(opplysninger))
                .build();
    }

    private List<Sporsmal> andreInntekterSporsmal(JsonOkonomiopplysninger opplysninger) {
        var harUtfyltAndreInntekterSporsmal = harBekreftelse(opplysninger, BEKREFTELSE_UTBETALING);
        var harSvartJaAndreInntekter = harBekreftelseTrue(opplysninger, BEKREFTELSE_UTBETALING);
        var utbetalingTyper = Arrays.asList(UTBETALING_UTBYTTE, UTBETALING_SALG, UTBETALING_FORSIKRING, UTBETALING_ANNET);
        var sporsmal = new ArrayList<Sporsmal>();
        sporsmal.add(
                new Sporsmal.Builder()
                        .withTittel("inntekt.inntekter.sporsmal")
                        .withErUtfylt(harUtfyltAndreInntekterSporsmal)
                        .withFelt(harUtfyltAndreInntekterSporsmal ?
                                booleanVerdiFelt(harSvartJaAndreInntekter, "inntekt.inntekter.true", "inntekt.inntekter.false") :
                                null
                        )
                        .build()
        );

        if (harSvartJaAndreInntekter) {
            var harSvartHvaHarDuMottattSporsmal = opplysninger.getUtbetaling().stream().anyMatch(utbetaling -> utbetalingTyper.contains(utbetaling.getType()));
            sporsmal.add(
                    new Sporsmal.Builder()
                            .withTittel("inntekt.inntekter.true.type.sporsmal")
                            .withErUtfylt(harSvartHvaHarDuMottattSporsmal)
                            .withFelt(harSvartHvaHarDuMottattSporsmal ? andreinntekterFelter(opplysninger) : null)
                            .build()
            );

            if (sporsmal.get(1).containsFeltWithSvar("json.okonomi.opplysninger.inntekt.inntekter.annet")) {
                var beskrivelseAvAnnet = opplysninger.getBeskrivelseAvAnnet();
                var harUtfyltAnnetFelt = beskrivelseAvAnnet != null && beskrivelseAvAnnet.getUtbetaling() != null && !beskrivelseAvAnnet.getUtbetaling().isBlank();
                sporsmal.add(
                        new Sporsmal.Builder()
                                .withTittel("inntekt.inntekter.true.type.annet")
                                .withErUtfylt(harUtfyltAnnetFelt)
                                .withFelt(harUtfyltAnnetFelt ?
                                        singletonList(new Felt.Builder().withType(Type.TEKST).withSvar(beskrivelseAvAnnet.getUtbetaling()).build()) :
                                        null
                                )
                                .build()
                );
            }
        }
        return sporsmal;
    }

    private List<Felt> andreinntekterFelter(JsonOkonomiopplysninger opplysninger) {
        var felter = new ArrayList<Felt>();
        addUtbetalingIfPresent(opplysninger, felter, UTBETALING_UTBYTTE, "json.okonomi.opplysninger.inntekt.inntekter.utbytte");
        addUtbetalingIfPresent(opplysninger, felter, UTBETALING_SALG, "json.okonomi.opplysninger.inntekt.inntekter.salg");
        addUtbetalingIfPresent(opplysninger, felter, UTBETALING_FORSIKRING, "json.okonomi.opplysninger.inntekt.inntekter.forsikringsutbetalinger");
        addUtbetalingIfPresent(opplysninger, felter, UTBETALING_ANNET, "json.okonomi.opplysninger.inntekt.inntekter.annet");
        return felter;
    }

    private void addUtbetalingIfPresent(JsonOkonomiopplysninger opplysninger, ArrayList<Felt> felter, String type, String key) {
        opplysninger.getUtbetaling().stream()
                .filter(utbetaling -> type.equals(utbetaling.getType()))
                .findFirst()
                .ifPresent(utbetaling -> felter.add(
                        new Felt.Builder()
                                .withType(Type.CHECKBOX)
                                .withSvar(key)
                                .build()
                ));
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/oppsummering/steg/inntektformue/AndreInntekter.kt
package no.nav.sosialhjelp.soknad.oppsummering.steg.inntektformue

import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.BEKREFTELSE_UTBETALING
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_ANNET
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_FORSIKRING
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_SALG
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_UTBYTTE
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger
import no.nav.sosialhjelp.soknad.oppsummering.dto.Avsnitt
import no.nav.sosialhjelp.soknad.oppsummering.dto.Felt
import no.nav.sosialhjelp.soknad.oppsummering.dto.Sporsmal
import no.nav.sosialhjelp.soknad.oppsummering.dto.SvarType
import no.nav.sosialhjelp.soknad.oppsummering.dto.Type
import no.nav.sosialhjelp.soknad.oppsummering.steg.StegUtils.booleanVerdiFelt
import no.nav.sosialhjelp.soknad.oppsummering.steg.StegUtils.createSvar
import no.nav.sosialhjelp.soknad.oppsummering.steg.inntektformue.InntektFormueUtils.harBekreftelse
import no.nav.sosialhjelp.soknad.oppsummering.steg.inntektformue.InntektFormueUtils.harBekreftelseTrue

class AndreInntekter {
    fun getAvsnitt(opplysninger: JsonOkonomiopplysninger): Avsnitt {
        return Avsnitt(
            tittel = "inntekt.inntekter.titel",
            sporsmal = andreInntekterSporsmal(opplysninger),
        )
    }

    private fun andreInntekterSporsmal(opplysninger: JsonOkonomiopplysninger): List<Sporsmal> {
        val harUtfyltAndreInntekterSporsmal = harBekreftelse(opplysninger, BEKREFTELSE_UTBETALING)
        val harSvartJaAndreInntekter = harBekreftelseTrue(opplysninger, BEKREFTELSE_UTBETALING)
        val utbetalingTyper = listOf(UTBETALING_UTBYTTE, UTBETALING_SALG, UTBETALING_FORSIKRING, UTBETALING_ANNET)
        val sporsmal = mutableListOf<Sporsmal>()
        sporsmal.add(
            Sporsmal(
                tittel = "inntekt.inntekter.sporsmal",
                erUtfylt = harUtfyltAndreInntekterSporsmal,
                felt =
                    if (harUtfyltAndreInntekterSporsmal) {
                        booleanVerdiFelt(
                            harSvartJaAndreInntekter,
                            "inntekt.inntekter.true",
                            "inntekt.inntekter.false",
                        )
                    } else {
                        null
                    },
            ),
        )
        if (harSvartJaAndreInntekter) {
            val harSvartHvaHarDuMottattSporsmal = opplysninger.utbetaling.any { utbetalingTyper.contains(it.type) }
            sporsmal.add(
                Sporsmal(
                    tittel = "inntekt.inntekter.true.type.sporsmal",
                    erUtfylt = harSvartHvaHarDuMottattSporsmal,
                    felt = if (harSvartHvaHarDuMottattSporsmal) andreinntekterFelter(opplysninger) else null,
                ),
            )
            if (sporsmal[1].containsFeltWithSvar("json.okonomi.opplysninger.inntekt.inntekter.annet")) {
                val beskrivelseAvAnnet = opplysninger.beskrivelseAvAnnet
                val harUtfyltAnnetFelt =
                    beskrivelseAvAnnet != null && beskrivelseAvAnnet.utbetaling != null && beskrivelseAvAnnet.utbetaling.isNotBlank()
                sporsmal.add(
                    Sporsmal(
                        tittel = "inntekt.inntekter.true.type.annet",
                        erUtfylt = harUtfyltAnnetFelt,
                        felt =
                            beskrivelseAvAnnet?.let {
                                if (harUtfyltAnnetFelt) {
                                    listOf(
                                        Felt(
                                            type = Type.TEKST,
                                            svar = createSvar(it.utbetaling, SvarType.TEKST),
                                        ),
                                    )
                                } else {
                                    null
                                }
                            },
                    ),
                )
            }
        }
        return sporsmal
    }

    private fun andreinntekterFelter(opplysninger: JsonOkonomiopplysninger): List<Felt> {
        val felter = mutableListOf<Felt>()
        addUtbetalingIfPresent(
            opplysninger,
            felter,
            UTBETALING_UTBYTTE,
            "json.okonomi.opplysninger.inntekt.inntekter.utbytte",
        )
        addUtbetalingIfPresent(
            opplysninger,
            felter,
            UTBETALING_SALG,
            "json.okonomi.opplysninger.inntekt.inntekter.salg",
        )
        addUtbetalingIfPresent(
            opplysninger,
            felter,
            UTBETALING_FORSIKRING,
            "json.okonomi.opplysninger.inntekt.inntekter.forsikringsutbetalinger",
        )
        addUtbetalingIfPresent(
            opplysninger,
            felter,
            UTBETALING_ANNET,
            "json.okonomi.opplysninger.inntekt.inntekter.annet",
        )
        return felter
    }

    private fun addUtbetalingIfPresent(
        opplysninger: JsonOkonomiopplysninger,
        felter: MutableList<Felt>,
        type: String,
        key: String,
    ) {
        opplysninger.utbetaling.firstOrNull { type == it.type }?.let {
            felter.add(
                Felt(
                    type = Type.CHECKBOX,
                    svar = createSvar(key, SvarType.LOCALE_TEKST),
                ),
            )
        }
    }
}
